<a href="https://colab.research.google.com/github/chris-kehl/Deep_Learning_with_Python/blob/main/Mnist_DeepLearning__Python_ch2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import tensorflow and the built-in dataset Mnist

In [16]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [3]:
train_images.shape

(60000, 28, 28)

View the shape of our taining data. We have 60000 images that are 28 x 28 pixels

In [17]:
len(train_labels)

60000

Our training labels are numbers that are single digit 0 - 9 dtype uint8

In [18]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

View the shape of our test data. We have 10000 images that are 28 x 28 pixels

In [19]:
test_images.shape

(10000, 28, 28)

In [20]:
len(test_labels)

10000

Our test lables are 0-9 single digit dtype uint8

In [21]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

The core of our neural networks are the layers. Here we import keras and keras layers. our layer will consist of a layer with 512 data points with the relu activation. The second layer is 10 data points with a softmax activation.

In [22]:
from tensorflow import keras
from tensorflow.keras import layers
model = keras.Sequential([
    layers.Dense(512, activation="relu"),
    layers.Dense(10, activation="softmax")
])

In [23]:
model.compile(optimizer="rmsprop",
              loss="sparse_categorical_crossentropy",
              metrics=["accuracy"])

The next process is to call our optimizer which is rmsprop, our loss function and the metrics we use for evaluation.

In [24]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

To get the images to properly work with our images we reshape them 28 * 28 = 784.
784 / 255 = 3.075 formatted to float32

In [25]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 2s 4ms/step - loss: 0.2567 - accuracy: 0.9258
Epoch 2/5
469/469 [==============================] - 2s 3ms/step - loss: 0.1032 - accuracy: 0.9693
Epoch 3/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0689 - accuracy: 0.9794
Epoch 4/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0502 - accuracy: 0.9848
Epoch 5/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0373 - accuracy: 0.9890


model.fit starts the trainig process for our training set, our batch_size is set to 128.

In [29]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0289 - accuracy: 0.9910
Epoch 2/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0221 - accuracy: 0.9938
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0166 - accuracy: 0.9950
Epoch 4/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0128 - accuracy: 0.9962
Epoch 5/5
469/469 [==============================] - 2s 3ms/step - loss: 0.0099 - accuracy: 0.9973


model.fit starts the testing process for our test set, our batch_size is set to 128.

In [30]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
predictions[0]

array([4.9861569e-13, 2.3384549e-16, 2.2509881e-10, 2.2385773e-07,
       5.6486973e-17, 7.3348229e-12, 4.6509207e-19, 9.9999976e-01,
       3.3295554e-11, 7.1796796e-10], dtype=float32)

Above we have 10 digits starting at 0 - 10. Our highest number is at the 7th position, so our first number [0] must be a 7.

In [31]:
predictions[0].argmax()

7

we call the prediction

In [32]:
predictions[0][7]

0.99999976

Our predicted number is a 7

In [33]:
test_labels[0]

7

We view the label and we can see the label is actually a 7

In [34]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"test_acc: {test_acc}")

313/313 [==============================] - 1s 3ms/step - loss: 0.0776 - accuracy: 0.9825
test_acc: 0.9825000166893005


Above we go through the entire dataset and test our dataset and achieve a 98.2% accuracy over the entire dataset.